### PR TITLE
enable Decimal support for Mysql

### PR DESCRIPTION
**Pull request by Fiedzia**
Whole decimal handling was really confusing, so I have no idea if I got this right, but seems to work.
I am still looking on how to write tests for it.

### CODE REVIEW

**Comment by killercup** [comment]
cc @rubdos who added BigDecimal support for Postgres

File: diesel/src/mysql/types/numeric.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    pub use self::bigdecimal::BigDecimal;
+
+    use types::{self, FromSql, ToSql, IsNull};
+
+
+    impl ToSql<types::Numeric, Mysql> for BigDecimal {
+        fn to_sql<W: Write>(&self, out: &mut W) -> Result<IsNull, Box<Error + Send + Sync>> {
+            // Mysql representats decimal type as char[]
+            out.write(&format!("{}", *self).into_bytes()).map(|_| IsNull::No).map_err(|e| e.into())
```
Comment (sgrif) [inline diesel/src/mysql/types/numeric.rs]: You need to use `write_all` here

Comment (sgrif) [inline diesel/src/mysql/types/numeric.rs]: How about `out.write_all(self.to_string().as_bytes())`?

Comment (sgrif) [inline diesel/src/mysql/types/numeric.rs]: Or even better: `out.write_fmt(format_args!("{}", self))`

Comment (sgrif) [inline diesel/src/mysql/types/numeric.rs]: How about `write!(out, "{}", self)`

File: diesel/src/mysql/types/numeric.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
+#[cfg(feature="bigdecimal")]
+pub mod bigdecimal {
+    extern crate num_traits;
+    extern crate num_bigint;
+    extern crate num_integer;
```
Comment (sgrif) [inline diesel/src/mysql/types/numeric.rs]: Do we need any of these besides `bigdecimal`?

File: diesel/src/mysql/types/mod.rs [hunk 3]
```diff
@@ -1,12 +1,19 @@
 #[cfg(feature = "chrono")]
 mod date_and_time;
+pub mod numeric;
 
 use byteorder::WriteBytesExt;
-use mysql::{Mysql, MysqlType};
+pub use mysql::{Mysql, MysqlType};
```
Comment (Eijebong) [inline diesel/src/mysql/types/mod.rs]: Why is this needed ?

File: diesel/src/mysql/types/mod.rs [hunk 4]
```diff
@@ -1,12 +1,19 @@
...
 use std::error::Error as StdError;
 use std::io::Write;
 use types::{ToSql, IsNull, FromSql, HasSqlType};
 
+pub mod sql_types {
```
Comment (Eijebong) [inline diesel/src/mysql/types/mod.rs]: We don't reexport `BigDecimal` for postgres, why should we for MySQL ?

I think you can remove the whole `sql_types` module :)

File: diesel/src/mysql/types/mod.rs [hunk 5]
```diff
@@ -58,3 +65,18 @@ impl HasSqlType<::types::Timestamp> for Mysql {
         MysqlType::Timestamp
     }
 }
+
```
Comment (Eijebong) [inline diesel/src/mysql/types/mod.rs]: ✂️

File: diesel/src/mysql/types/mod.rs [hunk 6]
```diff
@@ -58,3 +65,18 @@ impl HasSqlType<::types::Timestamp> for Mysql {
...
+
+impl HasSqlType<::types::Numeric> for Mysql {
+    fn metadata() -> MysqlType {
+        MysqlType::String
+    }
+}
+
```
Comment (Eijebong) [inline diesel/src/mysql/types/mod.rs]: ✂️

File: diesel/src/mysql/types/numeric.rs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    extern crate bigdecimal;
+
+    use std::error::Error;
+    use std::io::prelude::*;
+    use std::str::{self, FromStr};
+
+    use mysql::Mysql;
+
+    pub use self::bigdecimal::BigDecimal;
```
Comment (Eijebong) [inline diesel/src/mysql/types/numeric.rs]: Why does this need to be `pub use` ?

File: diesel/src/mysql/types/numeric.rs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+    use types::{self, FromSql, ToSql, IsNull};
+
```
Comment (Eijebong) [inline diesel/src/mysql/types/numeric.rs]: ✂️

**Comment by Fiedzia** [comment]
I've updated docs and I am not reexporting BigDecimal anymore. I'll will still need to add tests.

File: diesel/src/mysql/types/numeric.rs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    use mysql::{Mysql, MysqlType};
+
+    use self::bigdecimal::BigDecimal;
+
+    use types::{self, FromSql, ToSql, IsNull, HasSqlType};
+
+    impl ToSql<types::Numeric, Mysql> for BigDecimal {
+        fn to_sql<W: Write>(&self, out: &mut W) -> Result<IsNull, Box<Error + Send + Sync>> {
+            // Mysql representats decimal type as char[]
```
Comment (sgrif) [inline diesel/src/mysql/types/numeric.rs]: I don't think this comment is adding any value.

File: diesel/src/mysql/types/numeric.rs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+            write!(out, "{}", *self).map(|_| IsNull::No).map_err(|e| e.into())
+        }
+    }
+
+    impl FromSql<types::Numeric, Mysql> for BigDecimal {
+        fn from_sql(bytes: Option<&[u8]>) -> Result<Self, Box<Error+Send+Sync>> {
+            //_ numeric
```
Comment (sgrif) [inline diesel/src/mysql/types/numeric.rs]: I don't think this comment is adding any value.

File: diesel/src/mysql/types/numeric.rs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+            match str::from_utf8(not_none!(bytes)) {
```
Comment (sgrif) [inline diesel/src/mysql/types/numeric.rs]: Why do we need to do string conversion here? How about just `BigDecimal::parse_bytes`?

**Comment by Fiedzia** [comment]
I've enabled pg tests for mysql

**Comment by killercup** [comment]
@Fiedzia CI fails with

```
error[E0119]: conflicting implementations of trait `query_builder::query_id::QueryId` for type `types::Numeric`:
   --> /home/travis/build/diesel-rs/diesel/diesel/src/types/impls/mod.rs:160:9
```

Can you have a look at this?

**Comment by Eijebong** [comment]
Oh, you also need to rebase, it should be easy, you have to change the mysql metadata function so it takes `_: &()`

**Comment by sgrif** [comment]
Looks like something went wrong with the rebase (or with Github), this is showing all of #996 as part of this PR's changed files.

**Comment by Eijebong** [comment]
Hmm this is a git usage mistake. The first different commit is d9778d406b6db5a5d2ec1361c073021e0bacf9ea which is  
 9f4e9d379b6524a5c122979867df4a614b156f7a in this PR.

I don't know why, but the parent commit isn't the same.

@Fiedzia How did you rebase ? :p

**Comment by Fiedzia** [comment]
Its me. I did git pull --rebase upstream master

**Comment by Eijebong** [comment]
Mhhh weird that should have worked.

Did you have any conflict ? Did you resolve then properly ?

**Comment by Fiedzia** [comment]
I had conflicts and I resolved them with no particular issues.

**Comment by Fiedzia** [comment]
I believe I resolved all issues, builds are failing due to problem with rust nightly:
https://github.com/rust-lang/rust/issues/43153

File: diesel_tests/tests/types.rs [hunk 12]
```diff
@@ -463,6 +463,37 @@ fn pg_numeric_bigdecimal_to_sql() {
...
+        query_to_sql_equality::<Numeric, BigDecimal>(&expected, value)
+    }
+
+    quickcheck(correct_rep as fn(u64, u64) -> bool);
+
+    let test_values = vec![
+        "1.0",
+        "141.0",
+        "-1.0",
+        // Powers of 10k (numeric is represented in base 10k)
```
Comment (Eijebong) [inline diesel_tests/tests/types.rs]: This is not true for MySQL IIRC

File: diesel_tests/tests/types.rs [hunk 13]
```diff
@@ -488,6 +519,26 @@ fn pg_numeric_bigdecimal_from_sql() {
...
+    let expected_value: BigDecimal = "1.0".parse().expect("Could not parse to a BigDecimal");
+    assert_eq!(expected_value, query_single_value::<Numeric, BigDecimal>(query));
+
+    let query = "cast(141.00 as decimal)";
+    let expected_value: BigDecimal = "141.00".parse().expect("Could not parse to a BigDecimal");
+    assert_eq!(expected_value, query_single_value::<Numeric, BigDecimal>(query));
+
+    // Some non standard values:
+    let query = "cast(18446744073709551616 as decimal)"; // 2^64; doesn't fit in u64
+    //it is mysql, it will trim it even in strict mode
```
Comment (Eijebong) [inline diesel_tests/tests/types.rs]: Nit: "// It is"

**Comment by Fiedzia** [comment]
Linking the original issue here: #830

**Comment by Eijebong** [comment]
Thanks !
